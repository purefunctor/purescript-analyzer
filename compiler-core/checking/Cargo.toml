[package]
name = "checking"
version = "0.1.0"
edition = "2024"

[dependencies]
files = { version = "0.1.0", path = "../files" }
indexing = { version = "0.1.0", path = "../indexing" }
interner = { version = "0.1.0", path = "../interner" }
itertools = "0.14.0"
la-arena = "0.3.1"
lexing = { version = "0.1.0", path = "../lexing" }
lowering = { version = "0.1.0", path = "../lowering" }
parsing = { version = "0.1.0", path = "../parsing" }
rowan = "0.16.1"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }
