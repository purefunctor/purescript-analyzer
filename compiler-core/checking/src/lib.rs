pub mod check;
pub mod core;
pub mod debruijn;
