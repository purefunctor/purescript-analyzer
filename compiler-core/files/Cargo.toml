[package]
name = "files"
version = "0.1.0"
edition = "2024"

[dependencies]
indexmap = "2.7.1"
la-arena = "0.3.1"
rustc-hash = "2.1.1"
