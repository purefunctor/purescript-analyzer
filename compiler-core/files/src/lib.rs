use core::str;
use std::sync::Arc;

use indexmap::IndexMap;
use la_arena::{Idx, RawIdx};
use rustc_hash::FxBuildHasher;

#[derive(Debug, Default)]
pub struct Files {
    files: IndexMap<Arc<str>, Arc<str>, FxBuildHasher>,
}

pub struct File;

pub type FileId = Idx<File>;

impl Files {
    pub fn insert(&mut self, k: impl Into<Arc<str>>, v: impl Into<Arc<str>>) -> FileId {
        let k = k.into();
        let v = v.into();
        let (index, _) = self.files.insert_full(k, v);
        Idx::from_raw(RawIdx::from_u32(index as u32))
    }

    pub fn id(&self, k: &str) -> Option<FileId> {
        self.files.get_full(k).map(|(index, _, _)| Idx::from_raw(RawIdx::from_u32(index as u32)))
    }

    pub fn path(&self, id: FileId) -> Arc<str> {
        let index = id.into_raw().into_u32() as usize;
        let (path, _) =
            self.files.get_index(index).expect("invariant violated: expected valid FileId");
        Arc::clone(path)
    }

    pub fn content(&self, id: FileId) -> Arc<str> {
        let index = id.into_raw().into_u32() as usize;
        let (_, contents) =
            self.files.get_index(index).expect("invariant violated: expected valid FileId");
        Arc::clone(contents)
    }
}

#[cfg(test)]
mod tests {
    use super::Files;

    #[test]
    fn test_basic() {
        let mut files = Files::default();

        let k = "src/Main.purs";
        let v = "module Main where\n\n";

        let id = files.insert(k, v);

        assert_eq!(files.id(k), Some(id));
        assert_eq!(files.path(id).as_ref(), k);
        assert_eq!(files.content(id).as_ref(), v);
    }
}
