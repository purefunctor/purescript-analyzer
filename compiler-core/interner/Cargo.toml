[package]
name = "interner"
version = "0.1.0"
edition = "2024"

[dependencies]
hashbrown = "0.15.2"
la-arena = "0.3.1"
rustc-hash = "2.1.1"
