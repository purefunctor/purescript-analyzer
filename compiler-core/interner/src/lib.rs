use std::{
    hash::{BuildHasher, Hash},
    ops,
};

use hashbrown::{Equivalent, HashTable};
use la_arena::{Idx, RawIdx};
use rustc_hash::FxBuildHasher;

#[derive(Debug)]
pub struct Interner<T> {
    inner: Vec<T>,
    table: HashTable<usize>,
}

impl<T> Default for Interner<T> {
    fn default() -> Interner<T> {
        let inner = vec![];
        let table = HashTable::default();
        Interner { inner, table }
    }
}

impl<T: PartialEq> PartialEq for Interner<T> {
    fn eq(&self, other: &Interner<T>) -> bool {
        self.inner == other.inner
    }
}

impl<T: Eq> Eq for Interner<T> {}

impl<T: Eq + Hash> Interner<T> {
    pub fn intern(&mut self, value: T) -> Idx<T> {
        let hash = FxBuildHasher.hash_one(&value);
        let existing = self.table.find(hash, |&index| {
            let inner = &self.inner[index];
            &value == inner
        });
        let index = existing.copied().unwrap_or_else(|| {
            let index = self.inner.len();
            self.inner.push(value);
            self.table.insert_unique(hash, index, |&index| {
                let value = &self.inner[index];
                FxBuildHasher.hash_one(value)
            });
            index
        });
        Idx::from_raw(RawIdx::from_u32(index as u32))
    }

    pub fn get<Q>(&self, value: &Q) -> Option<Idx<T>>
    where
        Q: ?Sized + Hash + Equivalent<T>,
    {
        let hash = FxBuildHasher.hash_one(value);
        let index = *self.table.find(hash, |&index| {
            let inner = &self.inner[index];
            value.equivalent(inner)
        })?;
        Some(Idx::from_raw(RawIdx::from_u32(index as u32)))
    }
}

impl<T> ops::Index<Idx<T>> for Interner<T> {
    type Output = T;

    fn index(&self, index: Idx<T>) -> &Self::Output {
        let index = index.into_raw().into_u32() as usize;
        &self.inner[index]
    }
}

#[cfg(test)]
mod tests {
    use super::Interner;

    #[test]
    fn test_basic() {
        let mut interner = Interner::default();

        let hello_a = interner.intern("hello");
        let hello_b = interner.intern("hello");

        assert_eq!(hello_a, hello_b);
    }

    #[test]
    fn test_eq() {
        let mut interner_a = Interner::default();
        interner_a.intern("hello");
        interner_a.intern("world");

        let mut interner_b = Interner::default();
        interner_b.intern("hello");
        interner_b.intern("world");

        let mut interner_c = Interner::default();
        interner_c.intern("world");
        interner_c.intern("hello");

        assert_eq!(interner_a, interner_b);
        assert_ne!(interner_a, interner_c);
        assert_ne!(interner_b, interner_c);
    }
}
