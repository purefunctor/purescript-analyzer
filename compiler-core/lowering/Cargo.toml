[package]
name = "lowering"
version = "0.1.0"
edition = "2021"

[dependencies]
hashbrown = "0.15.2"
indexing = { version = "0.1.0", path = "../indexing" }
indexmap = "2.7.1"
itertools = "0.14.0"
la-arena = "0.3.1"
paste = "1.0.15"
rowan = "0.16.1"
rustc-hash = "2.1.1"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
insta = "1.42.1"
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
test_each_file = "0.3.4"
