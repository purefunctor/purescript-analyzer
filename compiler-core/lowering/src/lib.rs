mod algorithm;

pub mod intermediate;
pub mod scope;
pub mod source;

pub use intermediate::*;
pub use scope::*;
pub use source::*;

use indexing::FullIndexedModule;
use syntax::cst;

#[derive(Debug, PartialEq, Eq)]
pub struct FullLoweredModule {
    pub intermediate: Intermediate,
    pub source: LoweringSource,
    pub graph: LoweringGraph,
    pub graph_info: LoweringGraphInfo,
}

pub fn lower_module(module: &cst::Module, indexed: &FullIndexedModule) -> FullLoweredModule {
    let algorithm::State { intermediate, source, graph, graph_info, .. } =
        algorithm::lower_module(module, indexed);
    FullLoweredModule { intermediate, source, graph, graph_info }
}
