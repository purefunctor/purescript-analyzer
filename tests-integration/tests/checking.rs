use std::{fmt::Write, num::NonZeroU32, sync::Arc};

use analyzer::{QueryEngine, prim};
use checking::{
    check::{CheckContext, CheckState, PrimCore, unification::UnificationState},
    core::{Type, TypeId, TypeStorage, Variable, debruijn, pretty},
};
use files::{FileId, Files};
use interner::Interner;
use lowering::TypeVariableBindingId;

#[derive(Debug)]
struct InlineStorage {
    inner: Interner<checking::core::Type>,
}

impl Default for InlineStorage {
    fn default() -> Self {
        let inner = Interner::default();
        InlineStorage { inner }
    }
}

impl TypeStorage for InlineStorage {
    fn intern(&mut self, t: checking::core::Type) -> checking::core::TypeId {
        self.inner.intern(t)
    }

    fn index(&self, id: checking::core::TypeId) -> &checking::core::Type {
        &self.inner[id]
    }
}

struct TestEnv<'r> {
    engine: &'r QueryEngine,
    storage: InlineStorage,
    indexed: Arc<indexing::IndexedModule>,
    lowered: Arc<lowering::LoweredModule>,
    prim_indexed: Arc<indexing::IndexedModule>,
}

struct TestHandle<'r, 'w, S>
where
    S: TypeStorage,
{
    engine: &'r QueryEngine,
    state: &'w mut CheckState<'w, S>,
    context: &'r CheckContext<'r>,
}

impl<'r> TestEnv<'r> {
    fn new(engine: &'r QueryEngine, id: files::FileId) -> Self {
        let indexed = engine.indexed(id).unwrap();
        let lowered = engine.lowered(id).unwrap();
        let prim_indexed = {
            let prim_id = engine.prim_id();
            engine.indexed(prim_id).unwrap()
        };

        let storage = InlineStorage::default();

        Self { engine, storage, indexed, lowered, prim_indexed }
    }

    fn with_handle<F, R>(&mut self, f: F) -> R
    where
        F: FnOnce(TestHandle<InlineStorage>) -> R,
    {
        let mut state = CheckState::new(&mut self.storage);

        let prim = PrimCore::collect(self.engine, &mut state).unwrap();
        let context = CheckContext::new(prim, &self.indexed, &self.lowered, &self.prim_indexed);

        f(TestHandle { engine: self.engine, state: &mut state, context: &context })
    }
}

trait CheckStateExt {
    fn bound_variable(&mut self, index: u32) -> TypeId;

    fn function(&mut self, argument: TypeId, result: TypeId) -> TypeId;
}

impl<S: TypeStorage> CheckStateExt for CheckState<'_, S> {
    fn bound_variable(&mut self, index: u32) -> TypeId {
        let var = Variable::Bound(debruijn::Index(index));
        self.storage.intern(Type::Variable(var))
    }

    fn function(&mut self, argument: TypeId, result: TypeId) -> TypeId {
        self.storage.intern(Type::Function(argument, result))
    }
}

fn empty_engine() -> (QueryEngine, FileId) {
    let mut engine = QueryEngine::default();
    let mut files = Files::default();
    prim::configure(&mut engine, &mut files);

    let id = files.insert("Main.purs", "module Main where\n\n");
    let content = files.content(id);
    engine.set_content(id, content);

    (engine, id)
}

const FAKE_NONZERO_1: NonZeroU32 = NonZeroU32::new(1).unwrap();
const FAKE_NONZERO_2: NonZeroU32 = NonZeroU32::new(2).unwrap();

#[test]
fn test_solve_simple() {
    let (engine, id) = empty_engine();
    let mut env = TestEnv::new(&engine, id);

    env.with_handle(|TestHandle { engine, state, context }| {
        // [a :: Int, b :: String]
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);
        let codomain = state.bound.size();

        let unification = state.fresh_unification_type(context);
        let Type::Unification(unification_id) = *state.storage.index(unification) else {
            unreachable!("invariant violated");
        };

        state.solve(codomain, unification_id, context.prim.symbol).unwrap();

        let entry = state.unification.get(unification_id);
        let UnificationState::Solved(solution) = entry.state else {
            unreachable!("invariant violated");
        };

        let solution = pretty::print(engine, state, solution);
        let kind = pretty::print(engine, state, entry.kind);

        let snapshot = format!("{solution} :: {kind}");
        insta::assert_snapshot!(snapshot);
    });
}

#[test]
fn test_solve_bound() {
    let (engine, id) = empty_engine();
    let mut env = TestEnv::new(&engine, id);

    env.with_handle(|TestHandle { engine, state, context }| {
        // [a :: Int, b :: String]
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);
        let codomain = state.bound.size();

        let unification = state.fresh_unification_type(context);
        let Type::Unification(unification_id) = *state.storage.index(unification) else {
            unreachable!("invariant violated");
        };

        let bound_b = state.bound_variable(0);
        let bound_a = state.bound_variable(1);
        let b_to_a = state.function(bound_b, bound_a);

        state.solve(codomain, unification_id, b_to_a).unwrap();

        let entry = state.unification.get(unification_id);
        let UnificationState::Solved(solution) = entry.state else {
            unreachable!("invariant violated");
        };

        let solution = pretty::print(engine, state, solution);
        let kind = pretty::print(engine, state, entry.kind);

        let snapshot = format!("{solution} :: {kind}");
        insta::assert_snapshot!(snapshot);
    });
}

#[test]
fn test_solve_invalid() {
    let (engine, id) = empty_engine();
    let mut env = TestEnv::new(&engine, id);

    env.with_handle(|TestHandle { state, context, .. }| {
        // [a :: Int]
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);
        let codomain = state.bound.size();

        let unification = state.fresh_unification_type(context);
        let Type::Unification(unification_id) = *state.storage.index(unification) else {
            unreachable!("invariant violated");
        };

        // [a :: Int, b :: String]
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);

        let bound_b = state.bound_variable(0);
        let bound_a = state.bound_variable(1);
        let b_to_a = state.function(bound_b, bound_a);

        let solve_result = state.solve(codomain, unification_id, b_to_a);
        assert!(solve_result.is_none());
    })
}

#[test]
fn test_solve_promotion() {
    let (engine, id) = empty_engine();
    let mut env = TestEnv::new(&engine, id);

    env.with_handle(|TestHandle { engine, state, context }| {
        // [a :: Int]
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);

        let unification_a = state.fresh_unification_type(context);
        let Type::Unification(unification_id) = *state.storage.index(unification_a) else {
            unreachable!("invariant violated");
        };

        // [a :: Int, b :: String]
        state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);
        let codomain = state.bound.size();

        let unification_a_b = state.fresh_unification_type(context);
        state.solve(codomain, unification_id, unification_a_b).unwrap();

        let mut snapshot = String::default();

        for (index, entry) in state.unification.iter().enumerate() {
            let UnificationState::Solved(solution) = entry.state else { continue };
            let domain = entry.domain;
            let solution = pretty::print(engine, state, solution);
            writeln!(snapshot, "?{index}[{domain}] := {solution}").unwrap();
        }

        insta::assert_snapshot!(snapshot);
    });
}
