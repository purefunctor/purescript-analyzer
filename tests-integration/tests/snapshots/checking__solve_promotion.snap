---
source: tests-integration/tests/checking.rs
expression: snapshot
---
?0[:1] := ?2[:1]
?1[:2] := ?2[:1]
